client/state/notices/middleware.js

```diff
@@ -357,10 +346,15 @@ export const handlers = {
  * Middleware
  */
 
-export default ( { dispatch, getState } ) => next => action => {
+export default store => next => action => {
+	const rv = next( action );
+
 	if ( ! get( action, 'meta.notices.skip' ) && handlers.hasOwnProperty( action.type ) ) {
-		handlers[ action.type ]( dispatch, action, getState );
+		const noticeAction = handlers[ action.type ]( action );
+		if ( noticeAction ) {
+			store.dispatch( noticeAction );
+		}
 	}
 
-	return next( action );
+	return rv;
```

Comment:
why are you assign `rv` to a variable instead of the `return next( action )`.  Also what does it stand for? maybe return value?

Reply:
Oh, sorry I forgot to mention this change in the PR description. It deserves an explanation.

The point is that the dispatch of the original action (e.g., `POST_DELETE_SUCCESS`) should finish and hit the reducer before the dispatch of the notice action (`NOTICE_CREATE`) starts. If you don't do that, you'll see an odd order in the action log: `NOTICE_CREATE` is the first, `POST_DELETE_SUCCESS` is the second. It should be the other way around.

That's why I call `next` before dispatching the notice action.

Popular generic middleware libraries like [redux-saga](https://github.com/redux-saga/redux-saga/blob/master/packages/core/src/internal/middleware.js#L42-L44) or [redux-observable](https://github.com/redux-observable/redux-observable/blob/master/src/createEpicMiddleware.js#L72-L74) do exactly the same thing.

We should also change the data-layer middleware at some point. I remember I've seen some tricky code that needs to work around the odd order of the dispatched actions: HTTP request issued first, reducer hit and `isRequesting` flag updated second.

Reply:
ah! thanks for the explanation that makes a lot of sense 👍